File: blind75/49_minimum_window_substring/min_window.go
package min_window

import "math"

// MinWindow finds the minimum window substring containing all characters from t.
// Uses sliding window with hash maps for O(m+n) time complexity.
func MinWindow(s string, t string) string {
	if len(s) == 0 || len(t) == 0 || len(s) < len(t) {
		return ""
	}

	need := make(map[byte]int)
	for i := 0; i < len(t); i++ {
		need[t[i]]++
	}

	have := make(map[byte]int)
	required := len(need)
	formed := 0

	left := 0
	minLen := math.MaxInt32
	minLeft := 0

	for right := 0; right < len(s); right++ {
		char := s[right]
		have[char]++

		if need[char] > 0 && have[char] == need[char] {
			formed++
		}

		for formed == required && left <= right {
			if right-left+1 < minLen {
				minLen = right - left + 1
				minLeft = left
			}

			leftChar := s[left]
			have[leftChar]--
			if need[leftChar] > 0 && have[leftChar] < need[leftChar] {
				formed--
			}
			left++
		}
	}

	if minLen == math.MaxInt32 {
		return ""
	}
	return s[minLeft : minLeft+minLen]
}

File: blind75/49_minimum_window_substring/min_window.py
from collections import Counter


def min_window(s: str, t: str) -> str:
    """
    Finds the minimum window substring containing all characters from t.
    Uses sliding window with hash maps for O(m+n) time complexity.

    Args:
        s: Source string to search in
        t: Target string with required characters

    Returns:
        Minimum window substring, or empty string if no valid window exists
    """
    if not s or not t or len(s) < len(t):
        return ""

    need = Counter(t)
    have = {}
    required = len(need)
    formed = 0

    left = 0
    min_len = float('inf')
    min_left = 0

    for right, char in enumerate(s):
        have[char] = have.get(char, 0) + 1

        if char in need and have[char] == need[char]:
            formed += 1

        while formed == required and left <= right:
            if right - left + 1 < min_len:
                min_len = right - left + 1
                min_left = left

            left_char = s[left]
            have[left_char] -= 1
            if left_char in need and have[left_char] < need[left_char]:
                formed -= 1
            left += 1

    if min_len == float('inf'):
        return ""
    return s[min_left:min_left + min_len]
